### ISSUE TITLE
[✨] support async component function in `core.d.ts` signature

### ISSUE THREAD

**Issue by revintec** [BuilderIO/qwik#2765]
### Is your feature request related to a problem?

currently code like `export default component$(async()=>...)` would compile and work fine
but vscode is reporting syntax error 
```
Argument of type '() => Promise<JSXNode<any> | null>' is not assignable to parameter of type 'OnRenderFn<{}>'.
  Type 'Promise<JSXNode<any> | null>' is missing the following properties from type 'JSXNode<any>': type, props, key
ts(2345)
```

### Describe the solution you'd like

can we modify `core.d.ts` to update the `component$` signature?

from `export declare type OnRenderFn<PROPS> = (props: PROPS) => JSXNode<any> | null`
to `export declare type OnRenderFn<PROPS> = (props: PROPS) => JSXNode<any> | null | Promise<JSXNode<any>|null>`

it's a small change but I don't know if it's acceptable/compatible/safe

### Describe alternatives you've considered

`export default component$(async()=><>{(async()=>...)()}</>)`

but that's ugly and verbose

**Comment by DustinJSilk** [comment]
You shouldnt be putting async code directly in a component function.
Try place your async tasks in a `useTask$` hook instead, and update a store/signal with the new state.

**Comment by revintec** [comment]
@DustinJSilk that sure is an alternative solution, but it still is more verbose
why not using async code? is it because of the latency to FCP or the inability to display `loading` status?

**Comment by revintec** [comment]
BTW accessing signal(`useSignal()`) value(`signal.value`) after `await` seems not working

**Comment by shairez** [comment]
Hi @revintec !

This is not a simple change
We used to have `async` for `component$` but we removed it

According to @manucorporat ,  adding `async await` ability to component functions messes up the execution context.

**Comment by manucorporat** [comment]
> BTW accessing signal(useSignal()) value(signal.value) after await seems not working

right! that's why we dont allow it. this is not a good idea, all this processing should be done inside a useTask!

**Comment by revintec** [comment]
thanks for the history, I didn't know that
however I've noticed that `useTask$(...)` do support async functions, and there're already vscode linters telling users that using `qwik` functions after `await` may not gonna work, is it possible to allow users continue using `async` in `component$(...)` thus simplifies their code, but if they use any `qwik` function we use (the already existing) linter to warn them?

use case that allows async:
```typescript
component$(async()=>{
  const somedata=await(await fetch(...)).json()
  return <>{JSON.stringify(somedata)}</>
})
```

if not, we have to add more boilerplate code, it's like callback hell all over again:
```typescript
interface UselessDefinition{dataRef:Signal<...>}
const uselessComponentToIsolateReendering=component$<UselessDefinition>(({dataRef})=>{
  return<>{dataRef.value?JSON.stringify(dataRef.value):"loading maybe?"}</>
})
component$(()=>{
  const signal=useSignal<...>()
  fetch(...).then((uselessName)=>{
    uselessName.json().then((uselessName2)=>{
        signal.value=uselessName2
    }).catch(...)
  }).catch(...)
  return <uselessComponentToIsolateReendering dataRef={signal}/>
})
```

**Comment by revintec** [comment]
about the execution context, I understand why the got messed up, and it seems we can restore them like this
```
component$(({restoreExecutionContext}=>{
  ...
  await ...
  restoreExecutionContext()
  ...
  qwikFunctionCall...
  return ...
})
```

but let's leave that to another PR :)

**Comment by gioboa** [comment]
We moved this issue to `qwik-evolution` repo to create a RFC discussion for this.
Here is our [Qwik RFC process](https://github.com/QwikDev/qwik-evolution?tab=readme-ov-file#qwik-rfc-process) thanks.
